debian/manpages/flam3-render.1
.TH FLAM3-RENDER 1 "June 17, 2005"
.SH NAME
flam3-render \- renderer of .flam3 files
.SH DESCRIPTION
flam3-render comes with the electricsheep(1) package.
It takes a .flam3 file on stdin and renders it.
.SH OPTIONS
This program gets its opions through environment variables.
See flam3-render -h for details.
.br
.SH AUTHOR
This manual page was written by Alexis Sukrieh <sukria@sukria.net> 
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
electricsheep (2.6.8-5) unstable; urgency=low

  * Removed the debian revision from the version string so sheeps can be
    downloaded from the server.
    (closes: #391990)

 -- Alexis Sukrieh <sukria@debian.org>  Wed, 11 Oct 2006 07:54:36 +0200

electricsheep (2.6.8-4) unstable; urgency=high

  [ Jérémy Bobbio ]

  * debian/patches/04_Composite.dpatch
    + tells mpeg2dec_onroot to skip ARGB visuals
    when trying to find the best visual. Fix an issue when running inside a
    Composite-enabled X environment.
    (closes: #390902)

  [ Alexis Sukrieh ]

  * urgency set to high, for closing a RC bug in testing (BSP).

 -- Alexis Sukrieh <sukria@debian.org>  Sat,  7 Oct 2006 16:49:00 +0200

electricsheep (2.6.8-3) unstable; urgency=low

  * mpeg2dec/libmpeg2/idct_altivec.c
    Applied patch from p2@mind.be for fixing the FTBFS issue on ppc
    architectures.
    (closes: #386673)

 -- Alexis Sukrieh <sukria@debian.org>  Mon, 18 Sep 2006 08:50:14 +0200

electricsheep (2.6.8-2) unstable; urgency=low

  [ The "Hey babe, I'm a Debian Developer!" upload, part #1 ] 
  
  * Changed Maintainer field in debian/control to `sukria@debian.org'. 

 -- Alexis Sukrieh <sukria@debian.org>  Sat,  9 Sep 2006 10:27:36 +0200

electricsheep (2.6.8-1) unstable; urgency=low

  * New upstream release.
    (closes: 365652)

 -- Alexis Sukrieh <sukria@sukria.net>  Sat,  5 Aug 2006 10:43:00 +0200

electricsheep (2.6.6-1) unstable; urgency=low

  * New upstream release.
  * Fixed some packaging issues reported by Philipp Kern.
  * Updated autotools scripts:
    + Added AM_MAINTAINER_MODE to configure.in
    + Customized Makefile.am so we can drop debian/patches/01_Makefile.dpatch
  * debian/rules now uses DESTDIR instead of PREFIX, cleaner (thanks to
    Philipp Kern).
  * Policy revision: 3.7.2

 -- Alexis Sukrieh <sukria@sukria.net>  Tue,  2 May 2006 00:28:02 +0200

electricsheep (2.6.4-2) unstable; urgency=low

  * Added missing build-dependencies: libx11-dev, libxext-dev, x-dev, libxt-dev 
    (closes: #344653)

 -- Alexis Sukrieh <sukria@sukria.net>  Sun, 25 Dec 2005 10:26:31 +0100

electricsheep (2.6.4-1) unstable; urgency=low

  * New upstream release.
    + possible to preview electricsheep in xscreensaver while running other
    instances (closes: #273984).
    + possible to run electricsheep in standalone mode, with a specific
    datadir (closes: #341004).
  * Removed debian/electricsheep.postinst, not needed anymore with the new
    upstream code tree.

 -- Alexis Sukrieh <sukria@sukria.net>  Wed, 21 Dec 2005 15:49:57 +0100

electricsheep (2.6.3+cvs20051206-1) unstable; urgency=low

  * New upstream release (CVS snapshot from 2005/12/06)
    + Fixes several buffer overflows
  * New patch for canceling the use of xsetbg for avoiding bug #325689.
  * New configure.in for using AM_MAINTAINER_MODE.

 -- Alexis Sukrieh <sukria@sukria.net>  Tue,  6 Dec 2005 10:42:23 +0100

electricsheep (2.6.3-2) unstable; urgency=low

  * Removed the dependency to xlibs-dev, added libxv-dev instead which is the
    only one needed here.
  * Remove stubs maintainer scripts (they came from the old maintenance and I
    forgot to remove them, let's do a bit of cleaning).
    - debian/prerem
    - debian/postrm
  * Remove debian/REJECT (was containing the licence).
  * Remove the overkilling dependency xlibs-dev, the other build-dependencies
    are enough.
  * Changed the autotools scripts, using AM_MAINTAINER_MODE, added a
    bootstrap.
  * Removed duplicate debian/electricsheep.desktop file, is provided in
    debian/default-files.

 -- Alexis Sukrieh <sukria@sukria.net>  Tue, 29 Nov 2005 21:00:56 +0100

electricsheep (2.6.3-1) unstable; urgency=low

  * New upstream release.
    (closes: #318003)
  * New patch for Makefile.in.
  * Added a .desktop file for KDE (Thanks to Gonéri Le Bouder).
    (closes: #315420)

 -- Alexis Sukrieh <sukria@sukria.net>  Mon, 25 Jul 2005 14:12:57 +0200

electricsheep (2.6.2-1) unstable; urgency=low

  * New upstream release: 2.6.2.
    (closes: #309852)
  * Added myself as maintainer (Sander Smeenk offered it to me).
  * Added dpatch for handling changes applied on upstream sources.
  * Added all missing manpages.

 -- Alexis Sukrieh <sukria@sukria.net>  Fri, 17 Jun 2005 18:28:17 +0200

electricsheep (2.5-2) unstable; urgency=low

  + Fixed builddepends on libpng (Closes: #245816)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 25 Apr 2004 17:46:57 +0200

electricsheep (2.5-1) unstable; urgency=low

  + New Upstream Release (Closes: #243760)
  + vroot.h was updated which Closes: #217492
  
 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 18 Apr 2004 15:06:34 +0200

electricsheep (2.4-3) unstable; urgency=low
  
  + Fixed dephelper depends (Closes: #216014)
  + Fixed conflicts with libmpeg2-3-dev (Closes: #207342)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 19 Oct 2003 15:59:02 +0200

electricsheep (2.4-2) unstable; urgency=low

  + Correct dependencies (Closes: #206346)

 -- Sander Smeenk <ssmeenk@debian.org>  Sun, 24 Aug 2003 12:31:00 +0200
   
electricsheep (2.4-1) unstable; urgency=low

  + New upstream version (Closes: #203965)

 -- Sander Smeenk <ssmeenk@debian.org>  Thu, 14 Aug 2003 10:18:40 +0200

electricsheep (2.3-2) unstable; urgency=low

  + Moved *tif out of /usr/share to /usr/share/electricsheep.
  + Forgot to close bugs with my previous upload.
    Closes: #192510
  + Buildproblems on noni386 are fixed upstream
    Thanks Julien Lemoine for your effort.
    Closes: #167963, #182510
  + Added build dependency to libexpat1-dev (Closes: #185390)
  + Actually using CFLAGS from debian/rules now

 -- Sander Smeenk <ssmeenk@debian.org>  Tue, 18 Mar 2003 14:44:08 +0100

electricsheep (2.3-1) unstable; urgency=low

  + New upstream.
  + New debian/ structure, new Standards version, etc.

 -- Sander Smeenk <ssmeenk@debian.org>  Wed, 26 Feb 2003 09:26:08 +0100

electricsheep (2.1-6) unstable; urgency=low

  + 'electricsheep.org' the domain that pointed to the official

flam3/flam3-render.c
/*
    flame - cosmic recursive fractal flames
    Copyright (C) 1992-2003  Scott Draves <source@flam3.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

static char *flam3_render_c_id =
"@(#) $Id: flam3-render.c,v 1.8 2005/07/20 06:06:16 spotspot Exp $";

#ifdef WIN32
#define WINVER 0x0500
#include <windows.h>
#endif

#include "private.h"
#include "img.h"


int calc_nstrips(flam3_frame *spec) {
  double mem_required;
  double mem_available;
  int nstrips;
  flam3_genome *cps = spec->genomes;
#ifdef WIN32
  MEMORYSTATUS stat;
  stat.dwLength = sizeof(stat);
  GlobalMemoryStatus(&stat);
  mem_available = (double)stat.dwTotalPhys;
#elif defined(_SC_PHYS_PAGES) && defined(_SC_PAGESIZE)
  mem_available =
      (double)sysconf(_SC_PHYS_PAGES) * sysconf(_SC_PAGESIZE);
#else
  fprintf(stderr, "warning: unable to determine physical memory.\n");
  mem_available = 2e9;
#endif
  mem_available *= 0.8;
  mem_required = flam3_render_memory_required(spec);
  if (mem_available >= mem_required) return 1;
  nstrips = (int) ceil(mem_required / mem_available);
  return nstrips;
}

int main(int argc, char **argv) {
   flam3_frame f;
   char *ai;
   flam3_genome *cps;
   int ncps;
   int i;
   unsigned char *image;
   FILE *fp;
   char *fname;
   int this_size, last_size = -1;
   int strip;
   double center_y, center_base;
   int nstrips;
   char *prefix = args("prefix", "");
   char *out = args("out", NULL);
   char *format = getenv("format");
   int verbose = argi("verbose", 0);
   int bits = argi("bits", 64);
   int seed = argi("seed", 0);
   char *inf = getenv("in");
   double qs = argf("qs", 1.0);
   double ss = argf("ss", 1.0);
   double pixel_aspect = argf("pixel_aspect", 1.0);
   FILE *in;
   double zoom_scale;
   int channels;
   int start_time = time(0);

   if (1 != argc) {
     puts(docstring);
     exit(0);
   }

   srandom(seed ? seed : (time(0) + getpid()));

   fname = malloc(strlen(prefix) + 20);

   if (NULL == format) format = "jpg";
   if (strcmp(format, "jpg") &&
       strcmp(format, "ppm") &&
       strcmp(format, "png")) {
       fprintf(stderr,
	       "format must be either jpg, ppm, or png, not %s.\n",
	       format);
       exit(1);
   }

   channels = strcmp(format, "png") ? 3 : 4;
   
   if (pixel_aspect <= 0.0) {
     fprintf(stderr, "pixel aspect ratio must be positive, not %g.\n",
	     pixel_aspect);
     exit(1);
   }

   if (inf)
     in = fopen(inf, "rb");
   else
     in = stdin;
   if (NULL == in) {
     perror(inf);
     exit(1);
   }

   cps = flam3_parse_from_file(in, &ncps);
   if (NULL == cps) {
     exit(1);
   }
   for (i = 0; i < ncps; i++) {
     cps[i].sample_density *= qs;
     cps[i].height = (int)(cps[i].height * ss);
     cps[i].width = (int)(cps[i].width * ss);
     cps[i].pixels_per_unit *= ss;
   }

   if (out && (ncps > 1)) {
     fprintf(stderr, "hqi-flame: warning: writing multiple images "
	     "to one file.  all but last will be lost.\n");
   }


   for (i = 0; i < ncps; i++) {
       int real_height;

     if (verbose && ncps > 1) {
       fprintf(stderr, "flame = %d/%d ", i+1, ncps);
     }

      f.temporal_filter_radius = 0.0;
      f.genomes = &cps[i];
      f.ngenomes = 1;
      f.verbose = verbose;
      f.bits = bits;
      f.time = 0.0;
      f.pixel_aspect_ratio = pixel_aspect;
      f.progress = 0;

      if (getenv("nstrips")) {
	nstrips = atoi(getenv("nstrips"));
      } else {
	nstrips = calc_nstrips(&f);
      }

      if (nstrips > cps[i].height) {
	  fprintf(stderr, "cannot have more strips than rows but %d>%d.\n",
		  nstrips, cps[i].height);
	  exit(1);
      }

      this_size = channels * cps[i].width * cps[i].height;
      if (this_size != last_size) {
	 if (last_size != -1)
	    free(image);
	 last_size = this_size;
	 image = (unsigned char *) malloc(this_size);
      } else {
	  memset(image, 0, this_size);
      }

      cps[i].sample_density *= nstrips;
      real_height = cps[i].height;
      cps[i].height = (int) ceil(cps[i].height / (double) nstrips);
      center_y = cps[i].center[1];
      zoom_scale = pow(2.0, cps[i].zoom);
      center_base = center_y - ((nstrips - 1) * cps[i].height) /
	(2 * cps[i].pixels_per_unit * zoom_scale);

      for (strip = 0; strip < nstrips; strip++) {
	  unsigned char *strip_start =
	      image + cps[i].height * strip * cps[i].width * channels;
	  cps[i].center[1] = center_base +
	      cps[i].height * (double) strip /
	      (cps[i].pixels_per_unit * zoom_scale);

	  if ((cps[i].height * (strip + 1)) > real_height) {
	      int oh = cps[i].height;
	      cps[i].height = real_height - oh * strip;
	      cps[i].center[1] -= 
		  (oh - cps[i].height) * 0.5 /
		  (cps[i].pixels_per_unit * zoom_scale);
	  }

	  if (verbose && nstrips > 1) {
	      fprintf(stderr, "strip = %d/%d\n", strip+1, nstrips);
	  }
	  if (verbose && (1 == nstrips) && (ncps > 1)) {
	      fprintf(stderr, "\n");
	  }

	  flam3_render(&f, strip_start, cps[i].width, flam3_field_both, channels);

	  if (NULL != out) {
	      fname = out;
	  } else {
	      sprintf(fname, "%s%04d.%s", prefix, i, format);
	  }
	  fp = fopen(fname, "wb");
	  if (NULL == fp) {
	      perror(fname);
	      exit(1);
	  }
	  /* and write it out */
	  if (verbose) {
	      fprintf(stderr, "writing...");
	  }
	  if (!strcmp(format, "png")) {
	      write_png(fp, image, cps[i].width, real_height);
	  } else if (!strcmp(format, "jpg")) {
	      write_jpeg(fp, image, cps[i].width, real_height);
	  } else {
	      fprintf(fp, "P6\n");
	      fprintf(fp, "%d %d\n255\n", cps[i].width, real_height);
	      fwrite(image, 1, this_size, fp);
	  }
	  fclose(fp);
      }

      /* restore the cps values to their original values */
      cps[i].sample_density /= nstrips;
      cps[i].height = real_height;
      cps[i].center[1] = center_y;

      if (verbose) {
	fprintf(stderr, "done.\n");
      }
   }
   if (verbose && ((ncps > 1) || (nstrips > 1))) {
     int total_time = time(0) - start_time;

     if (total_time > 100)
       fprintf(stderr, "total time = %.1f minutes", total_time / 60.0);
     
     fprintf(stderr, "total time = %d seconds\n", total_time);
   }
}
